src/charactersheet/viewmodels/character/stats/health.js

```diff
@@ -0,0 +1,334 @@
+                    // reduce temporary hit Points, and do not apply to damage.
+                    this.health().tempHitPoints(remainingTempHP);
+                    newDamage = currentDamage;
+                } else { // remainingTempHP is negative. Damage continues on
+                        // to hit points
+                    this.health().tempHitPoints(0);
+                    newDamage = currentDamage - remainingTempHP;
+                }
+            } else {
+                newDamage = currentDamage + currentValue;
+            }
+            if (newDamage > maxHitPoints) {
+                if (newDamage >= maxHitPoints * 2) {
+                    // Trigger Massive Damage.
+                    this.massiveDamageTaken(true);
+                }
+                newDamage = maxHitPoints;
+            }
+            this.health().damage(newDamage);
+        },
+        owner: this
+    });
+
+    // Hit Point Interaction ***************************************************
+    profileDidUpdate = async (profile) => {
+        this.profile().importValues(profile.exportValues());
+        this.calculateHitDice();
+    };
+
+    calculateHitDice = () => {
+        const level = this.profile().level();
+
+        this.hitDiceList(new Array(level));
+        // TODO: support multiple hit die types for multiclasses
+        for (var i = 0; i < level; i++) {
+            const isVisible = (level - this.hitDice().used()) > i;
+            this.hitDiceList()[i] = ko.observable(isVisible);
+        }
+        // Tell the view to render the list again.
+        this.hitDiceList.valueHasMutated();
+    };
+
+    toggleHitDie = async (isAvailable) => {
+        if (isAvailable) {
+            this.hitDice().used(this.hitDice().used() + 1);
+        } else {
+            this.hitDice().used(this.hitDice().used() - 1);
+        }
+        await this.saveHitDice();
+    }
+
+    saveHitDice = async () => {
+        await this.hitDice().save();
+        this.calculateHitDice();
+    }
+
+    /**
+     * Reset the hit dice to an unused state up to the floor of half of the
+     * character's level.
+     *
+     * This will be used primarily for long rest resets.
+     */
+    resetHitDice = async () => {
+        const level = this.profile().level();
+        const restoredHitDice = Math.floor(level / 2) < 1 ? 1 : Math.floor(level / 2);
+        const remainingHitDice = this.hitDice().used() - restoredHitDice;
+        if (remainingHitDice < 0) {
+            this.hitDice().used(0);
+        } else {
+            this.hitDice().used(remainingHitDice);
+        }
+        await this.saveHitDice();
+    };
+
+    /**
+     * Fired when a long rest notification is recieved.
+     * Resets health and hit dice.
+     */
+    resetOnLongRest = async () => {
+        this.health().damage(0);
+        this.health().tempHitPoints(0);
+        await this.health().save();
+        await this.resetHitDice();
+    };
+
+    validation = {
```

Comment:
Should these be in the health model?

Reply:
I left these here because they aren't fields for the health model, per se, they are a form that uses data to write new data to the health model.